Add Send/Sync tests for all async fns

Darksonn opened the pull request:
Inspired by #2373.

jebrosen commented on tokio/src/stream/empty.rs:
```diff
@@ -10,6 +10,8 @@ use core::task::{Context, Poll};
 pub struct Empty<T>(PhantomData<T>);
 
 impl<T> Unpin for Empty<T> {}
+unsafe impl<T> Send for Empty<T> {}
```
This manual impl *should* be unnecessary if `Empty<T>` held a `PhantomData<fn() -> T>` instead. (I think that's the right `PhantomData` for this case).

carllerche replied:
Yes, I believe you are correct. Changing the `PhantomData` should be enough and avoid the `unsafe` impl.

Darksonn replied:
It isn't possible to do that because function pointers in const fns are unstable, and the constructor is a const fn. I can use `std::iter::Empty<T>` instead though.

jebrosen commented on tokio/src/stream/pending.rs:
```diff
@@ -10,6 +10,8 @@ use core::task::{Context, Poll};
 pub struct Pending<T>(PhantomData<T>);
 
 impl<T> Unpin for Pending<T> {}
+unsafe impl<T> Send for Pending<T> {}
```
Same as `Empty` RE: the choice of `PhantomData`.

carllerche commented:
Thanks for doing this :+1: great addition.

Only requested change is @jebrosen's comment. After that, this should be good to go.

jebrosen commented:
I didn't realize that would cause problems, sorry - the upstream issue is https://github.com/rust-lang/rust/issues/69459.

Since `std::iter::Empty` only gained unconditional `Send` and `Sync` in rust `1.42.0`, I think it would be better to stick to the `unsafe impl` after all or to use the hack with an associated `const` like in the mentioned issue.

Darksonn commented:
Alright, I've reverted to an unsafe impl. Would you take another look? Note that I also changed the macro since your last review, as I found a simpler macro.
